[platform][marvell] Arm 32-bit Arch support changes

Sabareesh-Kumar-Anandan opened the pull request:
Signed-off-by: Sabareesh Kumar Anandan <sanandan@marvell.com>

<!--
Please make sure you've read and understood our contributing guidelines:
https://github.com/Azure/SONiC/blob/gh-pages/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx" or "resolves #xxxx"

Please provide the following information:
-->

**- Why I did it**
**- How I did it**
Added Arm 32-bit arch build fixes
Added marvell armhf platform specific changes

**- How to verify it**
Compiled for arm 32-bit sonic image and verified it

**- Which release branch to backport (provide reason below if selected)**

<!--
- Note we only backport fixes to a release branch, *not* features!
- Please also provide a reason for the backporting below.
- e.g.
- [x] 202006
-->

- [ ] 201811
- [ ] 201911
- [ ] 202006

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->


**- A picture of a cute animal (not mandatory but encouraged)**

lguohan commented on platform/marvell-armhf/platform.conf:
```diff
@@ -104,8 +143,8 @@ prepare_boot_menu() {
-    fw_setenv ${FW_ARG} linuxargs "net.ifnames=0 loopfstype=squashfs loop=$image_dir/$FILESYSTEM_SQUASHFS varlog_size=$VAR_LOG" > /dev/null
-    fw_setenv ${FW_ARG} linuxargs_old "net.ifnames=0 loopfstype=squashfs loop=$image_dir_old/$FILESYSTEM_SQUASHFS varlog_size=$VAR_LOG" > /dev/null
+    fw_setenv ${FW_ARG} linuxargs "net.ifnames=0 loopfstype=squashfs loop=$image_dir/$FILESYSTEM_SQUASHFS varlog_size=$VAR_LOG apparmor=0 " > /dev/null
+    fw_setenv ${FW_ARG} linuxargs_old "net.ifnames=0 loopfstype=squashfs loop=$image_dir_old/$FILESYSTEM_SQUASHFS varlog_size=$VAR_LOG apparmor=0 " > /dev/null
```
why disable apparmor?

antony-rheneus replied:
apparmor kills ntpd, so we have disabled apparmor as of now.

lguohan replied:
is this a arm specific problem? ntpd seems running fine on x86 platform.

antony-rheneus replied:
Yes, it seems to be ARM specific

lguohan replied:
does not feel right to disable apparmor.

Sabareesh-Kumar-Anandan replied:
enabled apparmor

lguohan commented on platform/marvell-armhf/sonic-platform-et6448m/et6448m_plt_setup.sh:
```diff
@@ -52,6 +52,9 @@ main()
     et6448m_profile
+
+    python /etc/entropy.py &
+    /bin/sh /etc/inband_mgmt.sh
```
what is this inband mgmt port?

antony-rheneus replied:
ET6448M board, down't have OOB (mgmt) port. 
To access management, we have enabled in-band management feature in one of the data port.

lguohan commented on platform/marvell-armhf/sonic-platform-et6448m/inband_mgmt.sh:
```diff
@@ -0,0 +1,37 @@
+           ifconfig eth0 down
+           systemctl restart networking
+       fi
+       sleep 120
```
such sleep concerns me. why do we need so long wait time.

antony-rheneus replied:
At any time user can restart syncd docker by any means, at that time this loop has to monitor the reincarnation of the data port used for in-band and then restart ifup scripts for management port.
Longer the delay lesser the cpu utilization.

lguohan replied:
what ip addr here? are we assume dhcp on eth0? why do we need to wait for ip address to be assign on eth0?

antony-rheneus replied:
In the same board, via Ethernet USB dongle, management interface can also be created. so one difference would be ip address would not be assigned for in-band interface as it would have been created after syncd  init.
Whereas for external eth0, ip address would be assigned statically/dhcp, and it should not be disturbed unnecessarily.

lguohan commented on platform/marvell-armhf/sonic-platform-et6448m/inband_mgmt.sh:
```diff
@@ -0,0 +1,37 @@
+     modprobe i2c_mv64xxx
+     modprobe i2c-dev
+     modprobe i2c_mux_gpio
+     sleep 60
```
such sleep concerns me. why do we need so long wait time.

antony-rheneus replied:
Here, the logic is to start a monitoring loop for the creation of in-band management port once data ports are initialized through sai. 
So sleep has been introduced to start later after syncd docker is initialized.

lguohan commented on platform/marvell-armhf/sonic-platform-et6448m/inband_mgmt.sh:
```diff
@@ -0,0 +1,37 @@
+       fi
+       sleep 120
+   else
+     sleep 3
```
why sleep 3 here?

antony-rheneus replied:
At any time user can restart syncd docker by any means, at that time this loop has to monitor the reincarnation of the data port used for in-band and then restart ifup scripts for management port.
Longer the delay lesser the cpu utilization.

lguohan commented on platform/marvell-armhf/sonic-platform-et6448m/et6448m_plt_setup.sh:
```diff
@@ -52,6 +52,9 @@ main()
     fw_uboot_env_cfg
     et6448m_profile
+
+    python /etc/entropy.py &
```
what does this entropy.py use for?

antony-rheneus replied:
In A385 CPU arch, random number generator is not supported in CPU soc. During bootup, systemd requires lot of random seeds, so this has been handled in SW by providing entropy from pseudo random generator.

lguohan commented on platform/marvell-armhf/sonic-platform-et6448m/entropy.py:
```diff
@@ -0,0 +1,17 @@
+#!/usr/bin/python3.5
```
are we using python3.5 here? I do not remember we are using python3.5 on master image.

antony-rheneus replied:
we will get back to on this regarding python version.

qiluo-msft commented on platform/marvell-armhf/platform.conf:
```diff
@@ -10,28 +10,60 @@ fdt_addr=0x1000000
+fi
+# armhf-marvell_et6448m_52x-r0  - Platform = Et6448M
+# armhf-nokia_ixs7215_52x-r0 - Platform = Nokia IPD6448M
+PLATFORM=`sed -n 's/onie_platform=\(.*\)/\1/p' $MACH_FILE`
```
>s/onie_platform=\(.*\)/\1/p [](start = 18, length = 27)

`s/^onie_platform=\(.*\)/\1/p`
Could you add `^` at the beginning of the regex?

Sabareesh-Kumar-Anandan replied:
done

qiluo-msft commented on platform/marvell-armhf/platform.conf:
```diff
@@ -68,6 +100,13 @@ prepare_boot_menu() {
         echo "Using pre-configured uboot env"
     fi
+    if [ "$PLATFORM" = "armhf-nokia_ixs7215_52x-r0" ]; then
+        #FW_ENV_DEFAULT='/dev/mtd1 0x0000000 0x10000'
```
>#FW_ENV_DEFAULT='/dev/mtd1 0x0000000 0x10000' [](start = 8, length = 45)

remove unused code?

Sabareesh-Kumar-Anandan replied:
done

qiluo-msft commented on platform/marvell-armhf/platform.conf:
```diff
@@ -149,9 +191,53 @@ create_ubi_partition() {
+    blk_dev="/dev/sda"
+    demo_part=$(sgdisk -p $blk_dev | grep -e "$demo_volume_label" -e "$legacy_volume_label" | awk '{print $1}')
+    # ONIE partition size 168MB
+    onie_part_size=168
```
>168 [](start = 19, length = 3)

How do you determine this number?

Sabareesh-Kumar-Anandan replied:
If board is from manufacturing without sonic partition, then skip onie partition size as per platform partition layout

qiluo-msft commented on platform/marvell-armhf/platform.conf:
```diff
@@ -149,9 +191,53 @@ create_ubi_partition() {
+        echo "Error: Unable to create partition $demo_part on $blk_dev"
+        exit 1
+    }
+    partprobe || true
```
Seems wrong indentation level.

Sabareesh-Kumar-Anandan replied:
done

qiluo-msft commented on platform/marvell-armhf/platform.conf:
```diff
@@ -149,9 +191,53 @@ create_ubi_partition() {
+    # Make filesystem
+    mkfs.ext4 -L $demo_volume_label /dev/$demo_dev
+
+}
```
remove extra empty line

Sabareesh-Kumar-Anandan replied:
done

qiluo-msft commented:
> 

Remove extra empty line

---
Refers to: platform/marvell-armhf/platform.conf:269 in 235eb09. [](commit_id = 235eb09d4ed6b589a84728d7c7f17aad2c72edc8, deletion_comment = False)

qiluo-msft commented on platform/marvell-armhf/sonic-platform-et6448m/entropy.py:
```diff
@@ -0,0 +1,17 @@
+#!/usr/bin/python
```
>python [](start = 11, length = 6)

Is this script compatible with python3?

Sabareesh-Kumar-Anandan replied:
yes

qiluo-msft commented on platform/marvell-armhf/sonic-platform-et6448m/entropy.py:
```diff
@@ -0,0 +1,17 @@
+            d = urnd.read(512)
+            t = struct.pack('ii', 4 * len(d), len(d)) + d
+            fcntl.ioctl(rnd, RNDADDENTROPY, t)
+    time.sleep(30)
```
>time.sleep(30) [](start = 4, length = 14)

why sleep?

Sabareesh-Kumar-Anandan replied:
In A385 CPU arch, random number generator is not supported in CPU soc. During bootup, systemd requires lot of random seeds, so this has been handled in SW by providing entropy from pseudo random generator. Added sleep 30 to not to hog the CPU

qiluo-msft commented on platform/marvell-armhf/sonic-platform-et6448m/et6448m_plt_setup.sh:
```diff
@@ -52,6 +52,9 @@ main()
     fw_uboot_env_cfg
     et6448m_profile
+
+    python /etc/entropy.py &
```
>& [](start = 27, length = 1)

why run in background?

Sabareesh-Kumar-Anandan replied:
periodically we need to update random

qiluo-msft commented on platform/marvell-armhf/sonic-platform-et6448m/inband_mgmt.sh:
```diff
@@ -0,0 +1,37 @@
+     modprobe i2c_mv64xxx
+     modprobe i2c-dev
+     modprobe i2c_mux_gpio
+     sleep 60
```
>sleep 60 [](start = 5, length = 8)

In general I am worried about the magic number sleep in all scripts. Do you want to achieve some collaboration between them? Is there any risk in future platforms?

Sabareesh-Kumar-Anandan replied:
There is no use to start inband mgmt script before entire system is ready. So we have added 60sec sleep
This is specific only to this platform. Not applicable for other platforms.

qiluo-msft replied:
Thank you for the info. So my concern is "entire system is ready" < 60s, any risk here? If risk happens, what is the impact?

---
In reply to: [526620478](https://github.com/Azure/sonic-buildimage/pull/5749#discussion_r526620478) [](ancestors = 526620478)

Sabareesh-Kumar-Anandan replied:
There is no risk in that case. In-band mgmt port will be brought up after that.

qiluo-msft commented on sonic-slave-buster/Dockerfile.j2:
```diff
@@ -1,5 +1,7 @@
 {%- if CONFIGURED_ARCH == "armhf" %}
+FROM multiarch/qemu-user-static:x86_64-arm-5.0.0-2 as qemu
 FROM multiarch/debian-debootstrap:armhf-buster
+COPY --from=qemu /usr/bin/qemu-arm-static /usr/bin
```
>qemu /usr/bin/qemu-arm-static [](start = 12, length = 29)

Why copy a binary from another image? Seems a hack.

antony-rheneus replied:
Not a hack, but to change the qemu version from 4 to 5, and qemu 4 has severe segfaults with golang

qiluo-msft replied:
Could you achieve this goal by installing qemu-user-static package? I see version 5.0 [https://packages.debian.org/buster-backports/qemu-user-static](https://packages.debian.org/buster-backports/qemu-user-static)

---
In reply to: [526621660](https://github.com/Azure/sonic-buildimage/pull/5749#discussion_r526621660) [](ancestors = 526621660)

Sabareesh-Kumar-Anandan replied:
Couldn't able to install qemu from debian buster. 
Error Logs:
Unpacking qemu-user-static (1:5.0-14~bpo10+1) ...
dpkg (subprocess): unable to execute rm command for cleanup (rm): Too many levels of symbolic links
dpkg: error while cleaning up:
rm command for cleanup subprocess returned error exit status 2
Could not exec dpkg!
E: Sub-process /usr/bin/dpkg returned an error code (100)
E: Problem executing scripts DPkg::Post-Invoke 'rm -f /var/cache/apt/archives/*.deb /var/cache/apt/archives/partial/*.deb /var/cache/apt/*.bin || true'
E: Sub-process returned an error code
The command '/bin/sh -c apt-get install -y qemu-user-static=1:5.0-14~bpo10+1' returned a non-zero code: 100

Sabareesh-Kumar-Anandan replied:
Couldnt able to install qemu from debian buster.
Error Logs:
Unpacking qemu-user-static (1:5.0-14~bpo10+1) ...
dpkg (subprocess): unable to execute rm command for cleanup (rm): Too many levels of symbolic links
dpkg: error while cleaning up:
rm command for cleanup subprocess returned error exit status 2
Could not exec dpkg!
E: Sub-process /usr/bin/dpkg returned an error code (100)
E: Problem executing scripts DPkg::Post-Invoke 'rm -f /var/cache/apt/archives/*.deb /var/cache/apt/archives/partial/*.deb /var/cache/apt/*.bin || true'
E: Sub-process returned an error code
The command '/bin/sh -c apt-get install -y qemu-user-static=1:5.0-14~bpo10+1' returned a non-zero code: 100

qiluo-msft replied:
Not familiar with this error message. I notice you are using `dpkg` directly. Did you try use `apt-get`. Please add something like below into `/etc/apt/sources.list`
```
deb [arch=amdhf] [http://debian-archive.trafficmanager.net/debian](http://debian-archive.trafficmanager.net/debian) buster-backports main
```
I don't have an armhf dev environment, so please test it first.

---
In reply to: [530707840](https://github.com/Azure/sonic-buildimage/pull/5749#discussion_r530707840) [](ancestors = 530707840,526621660)

Sabareesh-Kumar-Anandan replied:
yes, I have already done that.
_echo 'deb [arch=armhf] http://ftp.debian.org/debian buster-backports main' >> /etc/apt/sources.list
apt-get install -y qemu-user-static=5.0-14~ bpo10+1_

qiluo-msft replied:
Sorry, I see `deb ...` line, but I cannot find the second part.
```
apt-get install -y qemu-user-static=5.0-14~ bpo10+1
```

---
In reply to: [532104630](https://github.com/Azure/sonic-buildimage/pull/5749#discussion_r532104630) [](ancestors = 532104630,530707840,526621660)

Sabareesh-Kumar-Anandan replied:
I tried below lines and it was not working. got same errors as above. So, I didnt add this to the PR.
echo 'deb [arch=armhf] http://ftp.debian.org/debian buster-backports main' >> /etc/apt/sources.list
apt-get install -y qemu-user-static=5.0-14~ bpo10+1

Mrvl-BenjaminHuang replied:
Hi @qiluo-msft, @Sabareesh-Kumar-Anandan,

   We actually can't upgrade the qemu-user-static by doing _apt-get install -y qemu-user-static=5.0-14~ bpo10+1_. Because the docker sonic-slave-buster is already running in the emulated arm environment, if you try to install qemu-user-static, it will pull the qemu-user-static binaries from the Debian repo for the arm arch and these binaries are to emulate something else from the arm host arch.

   What we want is to update a qemu-arm-static binary emulating the arm client arch from the x86_64 host arch, copying the new version from multiarch/qemu-user-static:x86_64-arm-5.0.0-2 is a correct way to do this.

Thanks,
Benjamin

qiluo-msft commented on platform/marvell-armhf/platform.conf:
```diff
@@ -162,10 +246,17 @@ mount_partition() {
+    if [ "$PLATFORM" = "armhf-marvell_et6448m_52x-r0" ]; then
+        trap_push "${onie_bin} umount /dev/ubi0_0|| true"
+        mount -t ubifs /dev/ubi0_0 $demo_mnt || {
+            echo "mount -t ubifs /dev/ubi0_0 $demo_mnt Failed"
```
>Failed [](start = 55, length = 6)

If this is a failure, exit?

Sabareesh-Kumar-Anandan replied:
on failure, trap push will take care as script is enabled with -e

qiluo-msft replied:
`mount ... || { echo "..." }` will always succeed. I see your message said "Failed", so I guess you need `exit` after `echo`.

---
In reply to: [526622606](https://github.com/Azure/sonic-buildimage/pull/5749#discussion_r526622606) [](ancestors = 526622606)

Sabareesh-Kumar-Anandan replied:
added exit

qiluo-msft commented:
> 

Remove extra empty line

---
Refers to: platform/marvell-armhf/platform.conf:267 in a085fde. [](commit_id = a085fde980e0671820548242eceb7159bd479783, deletion_comment = False)

Sabareesh-Kumar-Anandan commented:
> > 
> 
> Remove extra empty line
> 
> Refers to: platform/marvell-armhf/platform.conf:267 in [a085fde](https://github.com/Azure/sonic-buildimage/commit/a085fde980e0671820548242eceb7159bd479783). [](commit_id = [a085fde](https://github.com/Azure/sonic-buildimage/commit/a085fde980e0671820548242eceb7159bd479783), deletion_comment = False)

removed it

Sabareesh-Kumar-Anandan commented:
Hi @lguohan,
We have answered/addressed all the comments. Do you have any more comments?
Also, Can you please retrigger the tests? Failures were not caused by this PR.

qiluo-msft reviewed (approved):
LGTM. Please also check with other reviewers.
